### PR TITLE
buildDotnetModule: use setup hooks

### PR DESCRIPTION
**Pull request by IvarWithoutBones**
###### Motivation for this change

As suggested by @jtojnar in #153364. See the commit descriptions.

This is heavily inspired by `pkgs/build-support/rust`, so thanks to everyone who worked on that!

cc @prusnak.

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/build-support/build-dotnet-module/hooks/dotnet-configure-hook.sh [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+dotnetConfigureHook() {
+    echo "Executing dotnetConfigureHook"
+
+    runHook preConfigure
+
+    if [ -z "${enableParallelBuilding-}" ]; then
+        parallelFlag="--disable-parallel"
+    fi
+
+    export HOME=$(mktemp -d)
```
Comment (IvarWithoutBones) [inline pkgs/build-support/build-dotnet-module/hooks/dotnet-configure-hook.sh:12]: I'm not quite sure what the best place to set this is, all hooks depend on it.

**Comment by aanderse** [comment]
Sorry @IvarWithoutBones I don't think I have anything to offer here :disappointed: I wonder if @jonringer is able/willing to review?

**Comment by jonringer** [comment]
you rang?

**Comment by jonringer** [comment]
I'm fine with using this, most other ecosystems (at least python, rust) do something similar.

**Review by jonringer — APPROVED** [review]
gave it a quick glance, and looks like idiomatic setuphook usage

```
https://github.com/NixOS/nixpkgs/pull/155257

15 packages built:
ArchiSteamFarm alttpr-opentracker baget btcpayserver depotdownloader discordchatexporter-cli jackett nbxplorer netcoredbg opentabletdriver osu-lazer pinta python-language-server ryujinx wasabibackend
```

File: pkgs/build-support/build-dotnet-module/hooks/dotnet-configure-hook.sh (hunk 1 above)
Comment (SuperSandro2000) [reply pkgs/build-support/build-dotnet-module/hooks/dotnet-configure-hook.sh:12]: I would just export it at the earliest possible point we need it. We don't need to create a giant mess of hooks.

File: pkgs/build-support/build-dotnet-module/hooks/dotnet-build-hook.sh [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    fi
+
+    for project in ${projectFile[@]} ${testProjectFile[@]}; do
+        env \
+            dotnet build "$project" \
+                -maxcpucount:$maxCpuFlag \
+                -p:BuildInParallel=$parallelBuildFlag \
+                -p:ContinuousIntegrationBuild=true \
+                -p:Deterministic=true \
+                -p:Version="${version}" \
```
Comment (SuperSandro2000) [inline pkgs/build-support/build-dotnet-module/hooks/dotnet-build-hook.sh:23]: What happens if just name is set and version is left empty?

File: pkgs/build-support/build-dotnet-module/hooks/dotnet-build-hook.sh [hunk 3]
```diff
@@ -0,0 +1,37 @@
+declare -a projectFile testProjectFile dotnetBuildFlags dotnetFlags
+
+dotnetBuildHook() {
+    echo "Executing dotnetBuildHook"
+
+    runHook preBuild
+
+    if [ ! -z "${enableParallelBuilding-}" ]; then
+        maxCpuFlag="$NIX_BUILD_CORES"
+        parallelBuildFlag="true"
```
Comment (SuperSandro2000) [inline pkgs/build-support/build-dotnet-module/hooks/dotnet-build-hook.sh:10]: Do we local those variables in other hooks?

File: pkgs/build-support/build-dotnet-module/hooks/dotnet-fixup-hook.sh [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+    echo "Installed wrapper to: "$2""
+}
+
+dotnetFixupHook() {
+    echo "Executing dotnetFixupPhase"
+
+    if [ ! -z "${executables-}" ]; then
+        for executable in $executables; do
+            execPath="$out/lib/${pname}/$executable"
```
Comment (SuperSandro2000) [inline pkgs/build-support/build-dotnet-module/hooks/dotnet-fixup-hook.sh:20]: I think there should be a way to override this if pname does not match the lib name.

File: pkgs/build-support/build-dotnet-module/hooks/dotnet-build-hook.sh (hunk 3 above)
Comment (IvarWithoutBones) [reply pkgs/build-support/build-dotnet-module/hooks/dotnet-build-hook.sh:10]: `$maxCpuFlag` is used in `dotnet-check-hook`, and the value stays the same. The second one only seems to be used while actually building.

File: pkgs/build-support/build-dotnet-module/hooks/dotnet-fixup-hook.sh (hunk 4 above)
Comment (IvarWithoutBones) [reply pkgs/build-support/build-dotnet-module/hooks/dotnet-fixup-hook.sh:20]: Good idea, that could be useful in scenarios where `$out/lib/$pname` is already used as well. I've added the `installPath` flag :+1:

**Comment by IvarWithoutBones** [comment]
Resolved merge conflicts :+1:

**Comment by IvarWithoutBones** [comment]
Is there anything blocking this? I'd love to get it merged, i have some changes (almost) ready that depend on it

**Comment by SuperSandro2000** [comment]
I don't really have the time to look into such a big diff right now.

**Review by winterqt — APPROVED** [review]
Diff LGTM.

**Comment by IvarWithoutBones** [comment]
Thanks!

File: pkgs/build-support/build-dotnet-module/hooks/dotnet-fixup-hook.sh (hunk 4 above)
Comment (lilyinstarlight) [reply pkgs/build-support/build-dotnet-module/hooks/dotnet-fixup-hook.sh]: So, um, `installPath` was added as a flag to `buildDotnetModule`, but as I learned today, it doesn't seem to be used anywhere or do anything 😅

I imagine it probably got lost during refactoring, but I've opened #251267 about it

**Pull request merged**